cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.util;
+
+import co.cask.cdap.data2.transaction.coprocessor.CacheSupplier;
+import com.google.common.base.Supplier;
+import com.google.common.util.concurrent.Service;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Implementation of {@link CacheSupplier} that reference counts the {@link Service} and manages the lifecycle of
+ * the {@link Service} instance.
+ *
+ * @param <T> type of {@link Service} that is reference counted
+ */
+public class ReferenceCountedSupplier<T extends Service> implements CacheSupplier<T> {
+  private static final Log LOG = LogFactory.getLog(ReferenceCountedSupplier.class);
+
+  private final AtomicReference<T> instance;
+  private final AtomicInteger refCount;
+  private final Object lock;
+  private final String instanceName;
+  private final Supplier<T> instanceSupplier;
+
+  public ReferenceCountedSupplier(String instanceName, Object lock, AtomicInteger refCount, AtomicReference<T> instance,
+                                  Supplier<T> instanceSupplier) {
+    this.instanceName = instanceName;
+    this.lock = lock;
+    this.refCount = refCount;
+    this.instance = instance;
+    this.instanceSupplier = instanceSupplier;
+  }
+
+  @Override
+  public T get() {
+    synchronized (lock) {
+      if (instance.get() == null) {
+        // Instance has not been instantiated
+        T serviceInstance = instanceSupplier.get();
+        instance.set(serviceInstance);
+        serviceInstance.startAndWait();
+      }
+      int newCount = refCount.incrementAndGet();
+      if (LOG.isDebugEnabled()) {
+        LOG.debug(String.format("Incrementing reference count for %s: %d", instanceName, newCount));
+      }
+      return instance.get();
+    }
+  }
+
+  @Override
+  public void release() {
+    synchronized (lock) {
+      int newCount = refCount.decrementAndGet();
+      if (LOG.isDebugEnabled()) {
+        LOG.debug(String.format("Decrementing reference count for %s: %d", instanceName, newCount));
+      }
+
+      if (newCount == 0) {
+        try {
+          Service serviceInstance = instance.get();
+          serviceInstance.stopAndWait();
```

Comment:
It would be good to instantiate two different suppliers at the same time in the test to check multiple supplier types work correctly